### PR TITLE
TSK-1646: Bugfix for creation / update of Task with invalid Attachment

### PR DESCRIPTION
**Pull request by tge20**
---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [x] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [x] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [x] PR fulfills the ticket
- [x] Edge cases and unwanted side effects are tested
- [x] Readability

### CODE REVIEW

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java [hunk 1]
```diff
@@ -76,18 +78,69 @@
...
+      for (Attachment attachment : attachments) {
+        AttachmentImpl attachmentImpl = (AttachmentImpl) attachment;
+        verifyAttachment(attachmentImpl);
+        initAttachment(attachmentImpl, task);
+
+        try {
+          attachmentMapper.insert(attachmentImpl);
+          if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(
+                "TaskService.updateTask() for TaskId={} INSERTED an Attachment={}.",
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java:111]: Shouldn't this be TaskService.createTask? Also, should we maybe abstract the actual insert into a separate method, just like verifyAttachment and initAttachment?

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java [hunk 2]
```diff
@@ -76,18 +78,69 @@
...
+  }
+
+  private void insertNewAttachmentsOnTaskUpdate(TaskImpl newTaskImpl, TaskImpl oldTaskImpl)
+      throws AttachmentPersistenceException, ClassificationNotFoundException {
+    List<String> oldAttachmentIds =
+        oldTaskImpl.getAttachments().stream()
+            .map(AttachmentSummary::getId)
+            .collect(Collectors.toList());
+
+    List<Attachment> newAttachments =
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java:133]: should we maybe abstract this filtering of the new Attachments into a separate method?

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java (hunk 1 above)
Comment (tge20) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java:111]: Yes, you're right. I'll fix this.

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java (hunk 2 above)
Comment (gitgoodjhe) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java:133]: I was thinking about line 128 until 136

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java [hunk 3]
```diff
@@ -76,18 +78,69 @@
...
+        newTaskImpl.getAttachments().stream()
+            .filter(a -> !oldAttachmentIds.contains(a.getId()))
+            .collect(Collectors.toList());
+
+    for (Attachment attachment : newAttachments) {
+      insertNewAttachmentOnTaskUpdate(newTaskImpl, attachment);
+    }
+  }
+
+  private void updateModifiedAttachmentsOnTaskUpdate(TaskImpl newTaskImpl, TaskImpl oldTaskImpl) {
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/AttachmentHandler.java:143]: please verify the classification of the updated attachments.

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java [hunk 4]
```diff
@@ -1436,53 +1436,45 @@ private boolean desiredCallbackStateCanBeSetForFoundSummary(
...
-    }
-    try {
-      attachmentHandler.insertNewAttachmentsOnTaskCreation(task);
-    } catch (AttachmentPersistenceException e) {
-      throw new SystemException(
-          "Internal error when trying to insert new Attachments on Task Creation.", e);
-    }
+    attachmentHandler.insertNewAttachmentsOnTaskCreation(task);
+
+    serviceLevelHandler.updatePrioPlannedDueOfTask(task, null, false);
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java:1477]: comment why this is here

File: lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java [hunk 5]
```diff
@@ -709,6 +711,93 @@ void testCreateTaskWithWorkbasketMarkedForDeletion() throws Exception {
...
+    ClassificationSummaryImpl classification = new ClassificationSummaryImpl();
+    classification.setKey("NOT_EXISTING");
+    attachment.setClassificationSummary(classification);
+
+    TaskImpl task = (TaskImpl) makeNewTask(taskService);
+    task.addAttachment(attachment);
+
+    assertThatThrownBy(() -> taskService.createTask(task))
+        .isInstanceOf(ClassificationNotFoundException.class);
+  }
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java:757]: This one for update aswell

File: lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java [hunk 6]
```diff
@@ -709,6 +711,93 @@ void testCreateTaskWithWorkbasketMarkedForDeletion() throws Exception {
...
+
+  @WithAccessId(user = "user-1-1")
+  @Test
+  void should_ThrowException_When_CreatingTaskWithInvalidClassificationOfAttachment()
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java:761]: classificationkey missing

File: lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java [hunk 7]
```diff
@@ -709,6 +711,93 @@ void testCreateTaskWithWorkbasketMarkedForDeletion() throws Exception {
...
+
+  @WithAccessId(user = "user-1-1")
+  @Test
+  void should_FetchClassificationOfAttachment_When_CreatingTaskWithIncompleteClassification()
+      throws Exception {
+    Attachment attachment = taskService.newAttachment();
+    attachment.setObjectReference(
+        createObjectReference("COMPANY_A", "SYSTEM_A", "INSTANCE_A", "VNR", "1234567"));
+    ClassificationSummaryImpl classification = new ClassificationSummaryImpl();
+    classification.setKey("T2000");
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/CreateTaskAccTest.java:785]: classificationService.newClassification("T2000");

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java [hunk 8]
```diff
@@ -498,4 +502,110 @@ void testAddCustomAttributeToAttachment() throws Exception {
...
+    task.addAttachment(attachment);
+    taskService.updateTask(task);
+
+    task.removeAttachment(attachment.getId());
+    attachment.setClassificationSummary(null);
+    task.addAttachment(attachment);
+
+    assertThatThrownBy(() -> taskService.updateTask(task))
+        .isInstanceOf(InvalidArgumentException.class)
+        .hasMessage("Classification of Attachment must not be null.");
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java:531]: implementation could be better

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java [hunk 9]
```diff
@@ -498,4 +502,110 @@ void testAddCustomAttributeToAttachment() throws Exception {
...
+    task.addAttachment(attachment);
+
+    assertThatThrownBy(() -> taskService.updateTask(task))
+        .isInstanceOf(InvalidArgumentException.class)
+        .hasMessage("ObjectReference of Attachment must not be null.");
+  }
+
+  @WithAccessId(user = "user-1-1")
+  @Test
+  void should_ThrowException_When_UpdatingTaskWithInvalidClassificationOfAttachment()
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java:568]: key missing

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java [hunk 10]
```diff
@@ -498,4 +502,110 @@ void testAddCustomAttributeToAttachment() throws Exception {
...
+    task.addAttachment(attachment);
+
+    assertThatThrownBy(() -> taskService.updateTask(task))
+        .isInstanceOf(InvalidArgumentException.class)
+        .hasMessageContaining("ClassificationKey of Attachment must not be empty.");
+  }
+
+  @WithAccessId(user = "user-1-1")
+  @Test
+  void should_ThrowException_When_UpdatingTaskWithNotExistingClassificationOfAttachment()
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/UpdateTaskAttachmentsAccTest.java:581]: key missing

**Review by mustaphazorgati — APPROVED** [review]

**Pull request merged**